larpandoracontent/LArShowerRefinement/ShowerStartFinderTool.cc

```diff
@@ -0,0 +1,663 @@
+ *
+ *  @brief  Implementation of the shower start finder tool class.
+ *
+ *  $Log: $
+ */
+
+#include "Pandora/AlgorithmHeaders.h"
+#include "Pandora/AlgorithmTool.h"
+
+#include "larpandoracontent/LArObjects/LArTwoDSlidingFitResult.h"
+#include "larpandoracontent/LArObjects/LArTwoDSlidingShowerFitResult.h"
+
+#include "larpandoracontent/LArHelpers/LArConnectionPathwayHelper.h"
+#include "larpandoracontent/LArHelpers/LArGeometryHelper.h"
+#include "larpandoracontent/LArHelpers/LArPfoHelper.h"
+
+#include "larpandoracontent/LArShowerRefinement/ShowerStartFinderTool.h"
+
+using namespace pandora;
+
+namespace lar_content
+{
+
+ShowerStartFinderTool::ShowerStartFinderTool() :
+    m_spineSlidingFitWindow(10),
+    m_longitudinalCoordinateBinSize(1.f),
+    m_nInitialEnergyBins(5),
+    m_minSigmaDeviation(1.f),
+    m_maxEdgeGap(3.f),
+    m_longitudinalShowerFraction(0.85f),
+    m_minShowerOpeningAngle(2.f),
+    m_molliereRadius(4.5f),
+    m_showerSlidingFitWindow(1000),
+    m_maxLayerSeparation(5)
+{
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+StatusCode ShowerStartFinderTool::Run(const ParticleFlowObject *const pShowerPfo, const CartesianVector &peakDirection,
+    const HitType hitType, const CaloHitList &showerSpineHitList, CartesianVector &showerStartPosition, CartesianVector &showerStartDirection)
+{
+    try
+    {
+        // Fit the shower spine
+        CartesianPointVector hitPositions;
+        for (const CaloHit *const pCaloHit : showerSpineHitList)
+            hitPositions.push_back(pCaloHit->GetPositionVector());
+
+        const TwoDSlidingFitResult spineTwoDSlidingFit(&hitPositions, m_spineSlidingFitWindow, LArGeometryHelper::GetWireZPitch(this->GetPandora()));
+
+        // First obtain the longitudinal position of spine hits
+        LongitudinalPositionMap longitudinalPositionMap;
+        this->ObtainLongitudinalDecomposition(spineTwoDSlidingFit, showerSpineHitList, longitudinalPositionMap);
+
+        // Obtain spine energy profile
+        EnergySpectrumMap energySpectrumMap;
+        this->GetEnergyDistribution(showerSpineHitList, longitudinalPositionMap, energySpectrumMap);
+
+        // Now find the shower start position/direction
+        const bool isEndDownstream(peakDirection.GetZ() > 0.f);
+
+        this->FindShowerStartAndDirection(pShowerPfo, hitType, spineTwoDSlidingFit, energySpectrumMap, showerSpineHitList, isEndDownstream,
+            showerStartPosition, showerStartDirection);
+    }
+    catch (...)
+    {
+        return STATUS_CODE_FAILURE;
+    }
+
+    return STATUS_CODE_SUCCESS;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void ShowerStartFinderTool::ObtainLongitudinalDecomposition(const TwoDSlidingFitResult &spineTwoDSlidingFit,
```

Comment:
agreed...

`essentially i'm trying to get the exact longitudinal projection going 'ALONG' the path, not just using the PCA axis. that's because i wanted to be super sensitive to kinks`